data/tutorials/tt_02_file_manipulation.md

```diff
@@ -9,123 +9,618 @@ date: 2021-05-27T21:07:30-00:00
+you must explicitly flush the `out_channel` using the flush function.
+
+Here's an example:
+
+```ocaml
+let oc = open_out "myfile.txt"
+output_string oc "Hello, world!"
+flush oc
+```
+
+### Remembering to Close Channels
+
+When you open a file, the operating system allocates a file descriptor for it.
+File descriptors are a limited resource, so it's important to release them when
+you're done using them by closing the file. Furthermore, not closing an
+`out_channel` might result in data loss, because some data might still be in the
+buffer and not yet written to the file.
+
+However, simply closing the file at the end of the function is not sufficient,
+because if an exception is raised before the function reaches the point where it
+closes the file, the file will remain open. You should therefore ensure that
+files are always closed, even if an exception is raised. One way to do this is
+to use a `try ... with` block:
+
+```ocaml
+let read_from_file filename =
+  let ic = In_channel.open_text filename in
   try
-    let line = input_line ic in
-    (* read line, discard \n *)
-    print_endline line;
-    (* write the result to stdout *)
-    flush stdout;
-    (* write on the underlying device now *)
-    close_in ic
-    (* close the input channel *)
-  with e ->
-    (* some unexpected exception occurs *)
-    close_in_noerr ic;
-    (* emergency closing *)
-    raise e
+    let content = In_channel.input_all ic in
+    In_channel.close ic;
+    content
+  with exn ->
+    In_channel.close ic;
+    raise exn
+```
 
-(* exit with error: files are closed but channels are not flushed *)
+In this function, if an exception is raised when calling `In_channel.input_all`,
+the file is closed before the exception is re-raised.
 
-(* normal exit: all channels are flushed and closed *)
+Alternatively, you can use the `In_channel.with_open_*` or
+`Out_channel.with_open_*` functions covered above, which will ensure that the
+channel is closed if an exception is raised. The example above can be written as
+below:
+
+```ocaml
+let read_from_file filename =
+  In_channel.with_open_text filename In_channel.input_all
 ```
 
-We can compile and run this example:
+### Avoiding Namespace Conflicts with the Unix Module
+
+The OCaml `Unix` module and the `Stdlib` module both provide access to standard
+file descriptors with the same names like `stdin`, `stdout`, and `stderr`
+leading to type errors. You can fully qualify the `Stdlib` module channels or
+open the `Stdlib` module and use the unqualified names.
 
-<!-- $MDX dir=examples -->
-```sh
-$ ocamlopt -o file_manip file_manip.ml
-$ ./file_manip
-Hello!
+Examples:
+
+```ocaml
+let () =
+  let oc = Stdlib.open_out "myfile.txt" in
+  output_string oc "Hello, world!";
+  Stdlib.flush oc;
+  Stdlib.close_out oc
+
+let () =
+  open Stdlib
+  let oc = open_out "myfile.txt" in
+  output_string oc "Hello, world!";
+  flush oc;
+  close_out oc
+```
+
+### Be Aware of File Truncation with `open_out` and `open_out_bin`
```

Comment:
This is covered partly in "File Permissions and Open Flags". Perhaps "Open Flags" could have its own sub section in "Advanced Topics" ?